### PR TITLE
certbot.interfaces: reintroduce empty zope interfaces

### LINKED ISSUE #9485: Third-party plugin breakages in Certbot 2.x
https://github.com/certbot/certbot/pull/9161 removed the old `IAuthenticator`, `IInstaller` etc classes from `certbot.interface`. We had a 14-month deprecation period for these classes and eventually removed them in the release of v2.0.0.

A number of third-party plugins no longer work on Certbot v2.0.0, due to them referencing the removed classes. We have received reports about this on the community forums, this issue tracker, and other project issue trackers.

```python
@zope.interface.implementer(interfaces.IAuthenticator)
@zope.interface.provider(interfaces.IPluginFactory)
class Authenticator(dns_common.DNSAuthenticator):
```

Fixing this in Certbot v1.18.0 and later is simple: removing the `zope` decorators. In the case of `dns_common.DNSAuthenticator`, the plugin is still registered as an authenticator due to the way the new `abc`-based classes work.

However, prior to Certbot v1.18.0, this results in the plugin no longer being registered:

    # certbot certonly -d example.com --dry-run --preferred-challenges dns -a certbot-dns-inwx:dns-inwx
    PluginEntryPoint#certbot-dns-inwx:dns-inwx does not provide IPluginFactory, skipping
    Saving debug log to /var/log/letsencrypt/letsencrypt.log
    Could not choose appropriate plugin: The requested certbot-dns-inwx:dns-inwx plugin does not appear to be installed
    The requested certbot-dns-inwx:dns-inwx plugin does not appear to be installed

Plugin authors are stuck on their ability to support both old Certbot and new Certbot. Some integration projects have pinned Certbot back to v1.32.0 or told users to install Certbot v1.32.0 instead of v2.x, in order to deal with this problem. 

I think we should try to solve this issue for all the `dns_common.DNSAuthenticator`-based `certbot-dns-*` projects that appear to have real users (around 12,000+ based on UA statistics from April), if it is straightforward to do so. 

My proposal is to reintroduce some of the removed `zope` classes, without reintroducing zope itself. That approach can be seen [here](https://github.com/certbot/certbot/compare/master...alexzorin:zope-shim). The effect is that it will perform the required `zope`-based class registration in Certbot < v1.18.0, and be a no-operation in newer versions.

One downside is that we cannot entirely fix the problem. I'm not sure whether we can do anything about plugins that inherit from `certbot.plugins.common.Plugin` and then do `zope`-based registration to `IAuthenticator` or `IInstaller`,  without us backtracking on the `zope` removal. Based on the stats, I think total breakage in that case is less than 700 users. This number excludes a bunch from the Nextcloud project who I think are not affected due to pinning Certbot in their snap (at least, I can't see any complaints about it).

### PR DESCRIPTION
**Pull request by alexzorin**
Fixes #9485.

### CODE REVIEW

**Comment by alexzorin** [comment]
*Proposed forum post per #9485*

---

This post is intended for authors of third-party Certbot plugins.

Certbot v1.18.0 introduced a new way for third-party plugins to register themselves to Certbot. Notably, the [`certbot.interfaces`](https://eff-certbot.readthedocs.io/en/stable/api/certbot.interfaces.html#certbot.interfaces.Plugin) `abc`-based classes `Plugin`, `Authenticator` and `Installer`. These are the recommended way to write and register Certbot plugins. 

With the release of Certbot v1.19.0, the old `zope`-based classes (notably `IPluginFactory`, `IPlugin`, `IAuthenticator`, `IInstaller`) for registering plugins were deprecated and subsequently removed in v2.0.0. 

**Fixing compatibility with existing DNS plugins in v2.1.0**
With the release of Certbot v2.0.0, some users may have noticed that their Certbot would crash with an error:

    $ certbot plugins
    An unexpected error occurred:
    AttributeError: module 'certbot.interfaces' has no attribute 'IAuthenticator'

This is due to having a plugin installed that still referenced the old `zope`-based classes.

We are re-adding empty versions these classes in Certbot v2.1.0 in order to fix compatibility with these third-party plugins. Any plugin which inherits from `certbot.plugins.dns_common.DNSAuthenticator`, depends on `zope.interface` and registers to the deprecated `zope`-based classes, should work again. ([example of such a plugin](https://github.com/tengattack/certbot-dns-aliyun/blob/b3472d98a09f4c7ccc8b177aad139f4f876cf9ff/certbot_dns_aliyun/dns_aliyun.py#L20-L22)).

**Supporting new and old versions of Certbot in a plugin**
Third-party plugins that wish to support both new (>= v2.0.0) and old (< v1.18.0) versions of Certbot need to register themselves to both the `abc` and `zope`-based plugin interfaces.

Since the `abc`-based classes do not exist prior to Certbot v1.18.0, there are a number of options to do this in a way that is compatible across all Certbot versions:

```python
import zope.interface

from certbot import interfaces
from certbot.plugins import common
from certbot.plugins import dns_common

# Each method of registering a plugin here will work in every Certbot version,
# except for v2.0.0, but will work in v2.1.0 and later.
#
# All methods require "zope.interface" to be installed, declare this as a dependency
# of your plugin.

# Method 1.
# abc-based registration for Certbot >= v1.18.0 by inheriting from .common.Plugin.
# zope-based registration for Certbot < v1.18.0.
@zope.interface.implementer(interfaces.IAuthenticator)
@zope.interface.provider(interfaces.IPluginFactory)
class Authenticator(common.Plugin):
    ...


# Method 1 (for installer plugins)
@zope.interface.implementer(interfaces.IInstaller)
@zope.interface.provider(interfaces.IPluginFactory)
class Installer(common.Plugin):
    ...


# Method 2.
# Similar to Method 1, inheriting from .dns_common.DNSAuthenticator is recommended
# for DNS authenticators.  
@zope.interface.implementer(interfaces.IAuthenticator)
@zope.interface.provider(interfaces.IPluginFactory)
class DNSAuthenticator(dns_common.DNSAuthenticator):
    ...


# Method 3.
# For authenticators or installers that do not want to inherit from
# .common.Plugin or .dns_common.DNAuthenticator.
#
# Using Method 1 or Method 2 will usually suffice, as they include
# common functionality useful to most plugin types.
def register_installer(cls):
    # abc-based registration for Certbot >= v1.18.0.
    try:
        interfaces.Installer.register(cls)
    except AttributeError:
        pass

    # zope-based registration for Certbot < v1.18.0.
    zope.interface.implementer(interfaces.IInstaller)(cls)
    zope.interface.implementer(interfaces.IPluginFactory)(cls)

    return cls 


@register_installer
class OtherInstaller:
    ...
```

**Comment by HLFH** [comment]
Thanks for this PR!

Can this be merged, and Certbot v2.1.0 released?

File: certbot/CHANGELOG.md [hunk 1]
```diff
@@ -14,7 +14,12 @@ Certbot adheres to [Semantic Versioning](https://semver.org/).
 
 ### Fixed
 
-*
+* `IPluginFactory`, `IPlugin`, `IAuthenticator` and `IInstaller` have been re-added to
+  `certbot.interfaces`.
+    - This is to fix compatibility with a number of third-party DNS plugins which may
+      have started erroring with `AttributeError` in Certbot v2.0.0.
+    - Plugin authors can find more information about Certbot 2.x compatibility
+      [here]().
```
Comment (bmw) [inline certbot/CHANGELOG.md:22]: Reminder for us to add a link here.

Do you think it's worth putting this on [the wiki](https://github.com/certbot/certbot/wiki) or [static docs](https://eff-certbot.readthedocs.io/en/stable/) and linking there to simplify this or do you think making the community forum post early is OK?

**Review by bmw — CHANGES_REQUESTED** [review]
From https://github.com/certbot/certbot/issues/9485#issuecomment-1332006395:

> The forum post is a bit wordy, if you can see where places I can cut it down, I'd be glad to shorten it.

It is, however, I feel like it's hard to shorten it without trying to hide the details of what's going on here which I personally don't think we should do. If you wanted, we could call more attention to the actionable steps by making them bold or restructure things a bit maybe by having something like separate "What to do" and "Explanation" sections, but I'm personally fine with the post as it is.

Other than that, I have two suggestions for the content of the post.

1. Based on my testing, I think `zope.interface.implementer(interfaces.IPluginFactory)(cls)` should be `zope.interface.provider(interfaces.IPluginFactory)(cls)`.
2. I think it may be nice to link to [abc](https://docs.python.org/3/library/abc.html) and [zope](https://zopeinterface.readthedocs.io/) docs on the first usage of these terms.

Otherwise, this LGTM. Thanks so much for working on this Alex.

File: certbot/CHANGELOG.md (hunk 1 above)
Comment (alexzorin) [reply certbot/CHANGELOG.md:22]: Wiki sounds like a good place to put it, actually! I didn't particularly want to post this on the forums.

**Review by bmw — APPROVED** [review]
